frontend/public/components/pod-logs.jsx

```diff
@@ -3,16 +3,37 @@ import * as React from 'react';
 
 import { Dropdown, ResourceLog, ResourceName } from './utils';
 
-// Component to container dropdown or conatiner name if only one container in pod.
-const ContainerDropdown = ({currentContainer, containers, kind, onChange}) => {
-  const resourceName = (container) => {
-    return <ResourceName name={container.name} kind={kind} />;
+const reduceContainerStatuses = (accumulator, container, order) => {
+  return {
+    ...accumulator,
+    [container.name]: { ...container, order }
   };
-  const dropdownItems = _.mapValues(containers, resourceName);
+};
+
+const containerLabel = (container) => {
+  return <ResourceName name={container.name} kind="Container" />;
+};
+
+const ContainerDropdown = ({currentKey, containers, initContainers = {}, onChange}) => {
+  let spacerBefore = new Set();
+  let headerBefore = {};
+  if (!_.isEmpty(initContainers)) {
```

Comment:
Curious if you can extract this into a function?

```JavaScrtipt 
const headerBefore = makeHeader(initContainers);  // or something.... 
```
Creating vars within an `if ()` block tends to indicate that this bit could be atomic.

Reply:
Yeah, that could work.

Reply:
I think I could create two local functions to get the spacer and header props to pass on to the dropdown. Like this:

``` jsx
const ContainerDropdown = ({currentKey = '', containers, initContainers = {}, onChange}) => {
  const getSpacer = () => {
    const spacerBefore = new Set();
    if(_.isEmpty(initContainers)){
      return spacerBefore;
    }
    const firstInitContainer =  _.find(initContainers, {order: 0});
    spacerBefore.add(firstInitContainer.name);
    return spacerBefore;
  }
  
  const getHeaders = () => {
    if(_.isEmpty(initContainers)) {
      return {};
    }
    const firstContainer = _.find(containers, { order: 0 });
    const firstInitContainer =  _.find(initContainers, { order: 0 });
    return {
      [firstContainer.name]: 'Containers',
      [firstInitContainer.name]: 'Init Containers'
    };
  }
  
  const spacerBefore = getSpacer(initContainers);
  const headerBefore = getHeaders(initContainers, containers);
  // ...
```

Reply:
I don't really want to put them at a higher scope because they would only be used in the context of this `ContainerDropdown` component.

Reply:
I don't have a problem with the fact that they would not be reused (still local to the file, not exported)... but that does seem heavy-handed, might not be a great refactor.

Reply:
I guess I'm thinking of `ContainerDropdown` in terms of a class. Maybe I should explicitly make it into a `React.PureComponent` instead of using the shorthand function. Then it would make more sense that those two functions would be local.

Reply:
Could go that route.  
I'm prob just arguing for the clarity of isolation:

```JavaScript 
const headerBefore = (container, initContainer) => {
  return initContainer
    ? {
      [container.name]: 'Containers',
      [initContainer.name]: 'Init Containers'
    } : {};
}

const spacerBefore = (container) => {
  let spacerBefore = new Set();
  if (container) {
    spacerBefore.add(container.name);
  }
  return spacerBefore;
}

const ContainerDropdown = ({currentKey, containers, initContainers = {}, onChange}) => {
  const firstInitContainer = _.find(initContainers, {order: 0});
  const firstAppContainer = _.find(containers, { order: 0 }); 
  const headerBefore = headerBefore(firstContainer, firstInitContainer);
  const spacerBefore = spacerBefore(firstInitContaier);
  const dropdownItems = _.mapValues(_.merge(containers, initContainers), containerLabel);
  const title = _.get(dropdownItems, currentKey);
  return <Dropdown
    className="btn-group"
    headerBefore={headerBefore}
    items={dropdownItems}
    spacerBefore={spacerBefore}
    title={title}
    onChange={onChange} />;
};
```
ie, clearly separate the logic for each variable set.  You could probably go the class route as well, if you prefer.

Reply:
Btw, do you have to `let` a `new Set()` or can you still `const` it?  Pondering if `set.add()` is actually considered a mutation that would cause it to blow up... I need to look this up.

Reply:
`set.add()` returns a new set. you can use `const`

Reply:
OK looking at the code you have: I don't think it will work because you're not using the return value from set.add

Reply:
@benjaminapetersen As I was refactoring I wrote something almost identical to that. I think I might use that, but wrap it in an actual class so that I can make those two functions local.

Reply:
Sounds good. (Though technically using a class means they will actually be exported as public methods, thats probably the MOST public way to go)

Reply:
FWIW, keeping them as inline functions is probably least performant, but I don't know if it matters.  Just tossing it out there:

```JavaScript 
// every time outerFunc is called, new memory will have to be allocated
// for each inner function to create a new JS scope, etc. Not a big deal 
// for a component that won't be used frequently...
// but calling an external function does not do this.
const outerFunc = () => {
   const innerFunc = () => {};
   const innerFunc2 = () => {};
}

```